File: web/src/services/api/response-interception.ts
import { apiClient, request } from "@/services/api/request";

export type ResponseInterceptionRule = {
    contains: string;
    replace: string;
};

export type ResponseInterceptionSetting = {
    enabled: boolean;
    rules: ResponseInterceptionRule[];
};

export function getAdminResponseInterceptionSetting() {
    return request<{ setting: ResponseInterceptionSetting }>(apiClient.get("/admin/settings/response-interception"));
}

export function updateAdminResponseInterceptionSetting(setting: ResponseInterceptionSetting) {
    return request<{ setting: ResponseInterceptionSetting }>(apiClient.patch("/admin/settings/response-interception", setting));
}

File: backend/internal/handler/response_interception.go
package handler

import (
	"net/http"

	"infinite-canvas/backend/internal/service"

	"github.com/gin-gonic/gin"
)

func RegisterResponseInterceptionRoutes(r *gin.RouterGroup, svc *service.Service) {
	r.GET("/admin/settings/response-interception", func(c *gin.Context) {
		user, err := currentUser(c, svc)
		if err != nil {
			failService(c, err)
			return
		}
		setting, err := svc.AdminResponseInterceptionSetting(user)
		if err != nil {
			failService(c, err)
			return
		}
		ok(c, gin.H{"setting": setting})
	})

	r.PATCH("/admin/settings/response-interception", func(c *gin.Context) {
		user, err := currentUser(c, svc)
		if err != nil {
			failService(c, err)
			return
		}
		c.Request.Body = http.MaxBytesReader(c.Writer, c.Request.Body, 64<<10)
		var value service.ResponseInterceptionSetting
		if err := c.ShouldBindJSON(&value); err != nil {
			fail(c, http.StatusBadRequest, err)
			return
		}
		setting, err := svc.UpdateResponseInterceptionSetting(user, value)
		if err != nil {
			failService(c, err)
			return
		}
		ok(c, gin.H{"setting": setting})
	})
}
